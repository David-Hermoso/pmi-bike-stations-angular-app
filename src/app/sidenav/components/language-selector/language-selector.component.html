<mat-form-field
  appearance="fill">
  <mat-label> {{ 'language' | translate }}</mat-label>
  <mat-select [formControl]="languageSelectorControl">
    <mat-option *ngFor="let language of languages" [value]="language.value">
      {{ language.title | translate }}
    </mat-option>
  </mat-select>
</mat-form-field>
