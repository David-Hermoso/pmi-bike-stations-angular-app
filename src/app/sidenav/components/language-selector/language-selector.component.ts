import {Component, Input} from '@angular/core';
import {FormControl} from '@angular/forms';
import {TranslateService} from '@ngx-translate/core';

@Component({
  selector: 'app-language-selector',
  templateUrl: './language-selector.component.html',
  styleUrls: ['./language-selector.component.scss']
})
export class LanguageSelectorComponent {
  public languageSelectorControl: FormControl = new FormControl('es');
  public languages: any[] = [
    {
      value: 'es',
      title: 'spanish'
    },
    {
      value: 'ca',
      title: 'catalan'
    },
    {
      value: 'en',
      title: 'english'
    },
    {
      value: 'de',
      title: 'deutsch'
    }
  ];

  constructor(private translateService: TranslateService) {
    const currentLanguage = localStorage.getItem('lang') || 'es';
    this.translateService.use(currentLanguage);
    this.languageSelectorControl.setValue(currentLanguage);

    this.languageSelectorControl.valueChanges.subscribe((selectedLanguage: string) => {
      this.changeLanguage(selectedLanguage);
    });
  }

  private changeLanguage(language: string): void {
    this.translateService.use(language);
    localStorage.setItem('lang', language);
  }
}
