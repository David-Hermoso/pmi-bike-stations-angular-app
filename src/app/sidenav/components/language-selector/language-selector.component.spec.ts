import {LanguageSelectorComponent} from './language-selector.component';

describe('LanguageSelectorComponent', () => {
  it('should set current language from local storage', () => {
    const getItemSpy = jest
      .spyOn(global.Storage.prototype, 'getItem')
      .mockImplementation((key) => 'en');

    const useMock = jest.fn();

    const translateServiceMock = {
      use: useMock
    };
    const component: LanguageSelectorComponent = new LanguageSelectorComponent(translateServiceMock as any);

    expect(useMock).toHaveBeenCalledWith('en');
  });

  it('should set current language to default language', () => {
    const getItemSpy = jest
      .spyOn(global.Storage.prototype, 'getItem')
      .mockImplementation((key) => null);

    const useMock = jest.fn();

    const translateServiceMock = {
      use: useMock
    };
    const component: LanguageSelectorComponent = new LanguageSelectorComponent(translateServiceMock as any);

    expect(useMock).toHaveBeenCalledWith('es');
  });

  it('should call localStorage.setItem', () => {
    const setItemMock = jest.fn();
    jest
      .spyOn(global.Storage.prototype, 'setItem')
      .mockImplementation(setItemMock);

    const useMock = jest.fn();

    const translateServiceMock = {
      use: useMock
    };
    const component: LanguageSelectorComponent = new LanguageSelectorComponent(translateServiceMock as any);

    component.languageSelectorControl.patchValue('ca');

    expect(setItemMock).toMatchSnapshot();
  });
});

