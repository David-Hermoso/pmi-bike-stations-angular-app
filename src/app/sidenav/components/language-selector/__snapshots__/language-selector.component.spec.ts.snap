// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageSelectorComponent should call localStorage.setItem 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "lang",
      "ca",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
